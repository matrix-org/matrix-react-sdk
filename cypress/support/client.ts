/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/// <reference types="cypress" />

import type {
    MatrixClient,
    Room,
    IContent,
    FileType,
    Upload,
    UploadOpts,
    ICreateRoomOpts,
    ISendEventResponse,
} from "matrix-js-sdk/src/matrix";
import Chainable = Cypress.Chainable;

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace Cypress {
        interface Chainable {
            /**
             * Returns the MatrixClient from the MatrixClientPeg
             */
            getClient(): Chainable<MatrixClient | undefined>;
            /**
             * Create a room with given options.
             * @param options the options to apply when creating the room
             * @return the ID of the newly created room
             */
            createRoom(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Create a space with given options.
             * @param options the options to apply when creating the space
             * @return the ID of the newly created space (room)
             */
            createSpace(options: ICreateRoomOpts): Chainable<string>;
            /**
             * Invites the given user to the given room.
             * @param roomId the id of the room to invite to
             * @param userId the id of the user to invite
             */
            inviteUser(roomId: string, userId: string): Chainable<{}>;
            /**
             * Sets account data for the user.
             * @param type The type of account data.
             * @param data The data to store.
             */
            setAccountData(type: string, data: object): Chainable<{}>;
            /**
             * @param {string} roomId
             * @param {string} threadId
             * @param {string} eventType
             * @param {Object} content
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            sendEvent(
                roomId: string,
                threadId: string | null,
                eventType: string,
                content: IContent,
            ): Chainable<ISendEventResponse>;
            /**
             * @param {string} name
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setDisplayName(name: string): Chainable<{}>;
            /**
             * @param {string} url
             * @param {module:client.callback} callback Optional.
             * @return {Promise} Resolves: {} an empty object.
             * @return {module:http-api.MatrixError} Rejects: with an error response.
             */
            setAvatarUrl(url: string): Chainable<{}>;
            /**
             * Upload a file to the media repository on the homeserver.
             *
             * @param {object} file The object to upload. On a browser, something that
             *   can be sent to XMLHttpRequest.send (typically a File).  Under node.js,
             *   a a Buffer, String or ReadStream.
             */
            uploadContent(file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>>;
            /**
             * Turn an MXC URL into an HTTP one. <strong>This method is experimental and
             * may change.</strong>
             * @param {string} mxcUrl The MXC URL
             * @param {Number} width The desired width of the thumbnail.
             * @param {Number} height The desired height of the thumbnail.
             * @param {string} resizeMethod The thumbnail resize method to use, either
             * "crop" or "scale".
             * @param {Boolean} allowDirectLinks If true, return any non-mxc URLs
             * directly. Fetching such URLs will leak information about the user to
             * anyone they share a room with. If false, will return null for such URLs.
             * @return {?string} the avatar URL or null.
             */
            mxcUrlToHttp(
                mxcUrl: string,
                width?: number,
                height?: number,
                resizeMethod?: string,
                allowDirectLinks?: boolean,
            ): string | null;
            /**
             * Gets the list of DMs with a given user
             * @param userId The ID of the user
             * @return the list of DMs with that user
             */
            getDmRooms(userId: string): Chainable<string[]>;
            /**
             * Joins the given room by alias or ID
             * @param roomIdOrAlias the id or alias of the room to join
             */
            joinRoom(roomIdOrAlias: string): Chainable<Room>;
        }
    }
}

Cypress.Commands.add("getClient", (): Chainable<MatrixClient | undefined> => {
    return cy.window({ log: false }).then((win) => win.mxMatrixClientPeg.matrixClient);
});

Cypress.Commands.add("getDmRooms", (userId: string): Chainable<string[]> => {
    return cy
        .getClient()
        .then((cli) => cli.getAccountData("m.direct")?.getContent<Record<string, string[]>>())
        .then((dmRoomMap) => dmRoomMap[userId] ?? []);
});

Cypress.Commands.add("createRoom", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.window({ log: false }).then(async (win) => {
        const cli = win.mxMatrixClientPeg.matrixClient;
        const resp = await cli.createRoom(options);
        const roomId = resp.room_id;

        if (!cli.getRoom(roomId)) {
            await new Promise<void>((resolve) => {
                const onRoom = (room: Room) => {
                    if (room.roomId === roomId) {
                        cli.off(win.matrixcs.ClientEvent.Room, onRoom);
                        resolve();
                    }
                };
                cli.on(win.matrixcs.ClientEvent.Room, onRoom);
            });
        }

        return roomId;
    });
});

Cypress.Commands.add("createSpace", (options: ICreateRoomOpts): Chainable<string> => {
    return cy.createRoom({
        ...options,
        creation_content: {
            type: "m.space",
        },
    });
});

Cypress.Commands.add("inviteUser", (roomId: string, userId: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        const res = await cli.invite(roomId, userId);
        Cypress.log({ name: "inviteUser", message: `sent invite in ${roomId} for ${userId}` });
        return res;
    });
});

Cypress.Commands.add("setAccountData", (type: string, data: object): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAccountData(type, data);
    });
});

Cypress.Commands.add(
    "sendEvent",
    (roomId: string, threadId: string | null, eventType: string, content: IContent): Chainable<ISendEventResponse> => {
        return cy.getClient().then(async (cli: MatrixClient) => {
            return cli.sendEvent(roomId, threadId, eventType, content);
        });
    },
);

Cypress.Commands.add("setDisplayName", (name: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setDisplayName(name);
    });
});

Cypress.Commands.add("uploadContent", (file: FileType, opts?: UploadOpts): Chainable<Awaited<Upload["promise"]>> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.uploadContent(file, opts);
    });
});

Cypress.Commands.add("setAvatarUrl", (url: string): Chainable<{}> => {
    return cy.getClient().then(async (cli: MatrixClient) => {
        return cli.setAvatarUrl(url);
    });
});

Cypress.Commands.add("joinRoom", (roomIdOrAlias: string): Chainable<Room> => {
    return cy.getClient().then((cli) => cli.joinRoom(roomIdOrAlias));
});
