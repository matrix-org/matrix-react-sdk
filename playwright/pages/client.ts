/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { JSHandle, Page } from "@playwright/test";
import { PageFunctionOn } from "playwright-core/types/structs";

import type {
    IContent,
    ICreateRoomOpts,
    ISendEventResponse,
    MatrixClient,
    Room,
    IRoomDirectoryOptions,
} from "matrix-js-sdk/src/matrix";

export class Client {
    protected client: JSHandle<MatrixClient>;

    protected getClientHandle(): Promise<JSHandle<MatrixClient>> {
        return this.page.evaluateHandle(() => window.mxMatrixClientPeg.get());
    }

    public async prepareClient(): Promise<JSHandle<MatrixClient>> {
        if (!this.client) {
            this.client = await this.getClientHandle();
        }
        return this.client;
    }

    public constructor(protected readonly page: Page) {
        page.on("framenavigated", async () => {
            this.client = null;
        });
    }

    public evaluate<R, Arg, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, Arg, R>,
        arg: Arg,
    ): Promise<R>;
    public evaluate<R, O extends MatrixClient = MatrixClient>(
        pageFunction: PageFunctionOn<O, void, R>,
        arg?: any,
    ): Promise<R>;
    public async evaluate<T>(fn: (client: MatrixClient) => T, arg?: any): Promise<T> {
        await this.prepareClient();
        return this.client.evaluate(fn, arg);
    }

    /**
     * @param roomId ID of the room to send the event into
     * @param threadId ID of the thread to send into or null for main timeline
     * @param eventType type of event to send
     * @param content the event content to send
     */
    public async sendEvent(
        roomId: string,
        threadId: string | null,
        eventType: string,
        content: IContent,
    ): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            async (client, { roomId, threadId, eventType, content }) => {
                return client.sendEvent(roomId, threadId, eventType, content);
            },
            { roomId, threadId, eventType, content },
        );
    }

    /**
     * Send a message as a bot into a room
     * @param roomId ID of the room to send the message into
     * @param content the event content to send
     */
    public async sendMessage(roomId: string, content: IContent): Promise<ISendEventResponse> {
        const client = await this.prepareClient();
        return client.evaluate(
            (client, { roomId, content }) => {
                return client.sendMessage(roomId, content);
            },
            {
                roomId,
                content,
            },
        );
    }

    /**
     * Create a room with given options.
     * @param options the options to apply when creating the room
     * @return the ID of the newly created room
     */
    public async createRoom(options: ICreateRoomOpts): Promise<string> {
        const client = await this.prepareClient();
        return await client.evaluate(async (cli, options) => {
            const resp = await cli.createRoom(options);
            const roomId = resp.room_id;
            if (!cli.getRoom(roomId)) {
                await new Promise<void>((resolve) => {
                    const onRoom = (room: Room) => {
                        if (room.roomId === roomId) {
                            cli.off(window.matrixcs.ClientEvent.Room, onRoom);
                            resolve();
                        }
                    };
                    cli.on(window.matrixcs.ClientEvent.Room, onRoom);
                });
            }
            return roomId;
        }, options);
    }

    /**
     * Create a space with given options.
     * @param options the options to apply when creating the space
     * @return the ID of the newly created space (room)
     */
    public async createSpace(options: ICreateRoomOpts): Promise<string> {
        return this.createRoom({
            ...options,
            creation_content: {
                ...options.creation_content,
                type: "m.space",
            },
        });
    }

    /**
     * Joins the given room by alias or ID
     * @param roomIdOrAlias the id or alias of the room to join
     */
    public async joinRoom(roomIdOrAlias: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(async (client, roomIdOrAlias) => {
            return await client.joinRoom(roomIdOrAlias);
        }, roomIdOrAlias);
    }

    /**
     * Make this bot join a room by name
     * @param roomName Name of the room to join
     */
    public async joinRoomByName(roomName: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate(
            (client, { roomName }) => {
                const room = client.getRooms().find((r) => r.getDefaultRoomName(client.getUserId()) === roomName);
                if (room) {
                    return client.joinRoom(room.roomId);
                }
                throw new Error(`Bot room join failed. Cannot find room '${roomName}'`);
            },
            {
                roomName,
            },
        );
    }

    /**
     * Invites the given user to the given room.
     * @param roomId the id of the room to invite to
     * @param userId the id of the user to invite
     */
    public async inviteUser(roomId: string, userId: string): Promise<void> {
        const client = await this.prepareClient();
        await client.evaluate((client, { roomId, userId }) => client.invite(roomId, userId), {
            roomId,
            userId,
        });
    }

    public async publicRooms(options?: IRoomDirectoryOptions): ReturnType<MatrixClient["publicRooms"]> {
        const client = await this.prepareClient();
        return await client.evaluate((client, options) => {
            return client.publicRooms(options);
        }, options);
    }
}
