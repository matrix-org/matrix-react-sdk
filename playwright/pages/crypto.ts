/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { APIRequestContext, Page, expect } from "@playwright/test";

import { HomeserverInstance } from "../plugins/homeserver";

export class Crypto {
    public constructor(
        private page: Page,
        private homeserver: HomeserverInstance,
        private request: APIRequestContext,
    ) {}

    /**
     * Check that the user has published cross-signing keys, and that the user's device has been cross-signed.
     */
    public async assertDeviceIsCrossSigned(): Promise<void> {
        const { userId, deviceId, accessToken } = await this.page.evaluate(() => ({
            userId: window.mxMatrixClientPeg.get().getUserId(),
            deviceId: window.mxMatrixClientPeg.get().getDeviceId(),
            accessToken: window.mxMatrixClientPeg.get().getAccessToken(),
        }));

        const res = await this.request.post(`${this.homeserver.config.baseUrl}/_matrix/client/v3/keys/query`, {
            headers: { Authorization: `Bearer ${accessToken}` },
            data: { device_keys: { [userId]: [] } },
        });
        const json = await res.json();

        // there should be three cross-signing keys
        expect(json.master_keys[userId]).toHaveProperty("keys");
        expect(json.self_signing_keys[userId]).toHaveProperty("keys");
        expect(json.user_signing_keys[userId]).toHaveProperty("keys");

        // and the device should be signed by the self-signing key
        const selfSigningKeyId = Object.keys(json.self_signing_keys[userId].keys)[0];

        expect(json.device_keys[userId][deviceId]).toBeDefined();

        const myDeviceSignatures = json.device_keys[userId][deviceId].signatures[userId];
        expect(myDeviceSignatures[selfSigningKeyId]).toBeDefined();
    }
}
