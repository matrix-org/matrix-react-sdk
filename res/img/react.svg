<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
