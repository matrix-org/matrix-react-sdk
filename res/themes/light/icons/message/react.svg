<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>