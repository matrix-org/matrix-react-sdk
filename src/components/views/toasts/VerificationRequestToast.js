/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import PropTypes from 'prop-types';
import * as sdk from "../../../index";
import { _t } from '../../../languageHandler';
import {MatrixClientPeg} from '../../../MatrixClientPeg';
import {RIGHT_PANEL_PHASES} from "../../../stores/RightPanelStorePhases";
import {userLabelForEventRoom} from "../../../utils/KeyVerificationStateObserver";
import dis from "../../../dispatcher";
import ToastStore from "../../../stores/ToastStore";
import Modal from "../../../Modal";

export default class VerificationRequestToast extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = {counter: Math.ceil(props.request.timeout / 1000)};
    }

    componentDidMount() {
        const {request} = this.props;
        if (request.timeout && request.timeout > 0) {
            this._intervalHandle = setInterval(() => {
                let {counter} = this.state;
                counter = Math.max(0, counter - 1);
                this.setState({counter});
            }, 1000);
        }
        request.on("change", this._checkRequestIsPending);
        // We should probably have a separate class managing the active verification toasts,
        // rather than monitoring this in the toast component itself, since we'll get problems
        // like the toasdt not going away when the verification is cancelled unless it's the
        // one on the top (ie. the one that's mounted).
        // As a quick & dirty fix, check the toast is still relevant when it mounts (this prevents
        // a toast hanging around after logging in if you did a verification as part of login).
        this._checkRequestIsPending();
    }

    componentWillUnmount() {
        clearInterval(this._intervalHandle);
        const {request} = this.props;
        request.off("change", this._checkRequestIsPending);
    }

    _checkRequestIsPending = () => {
        const {request} = this.props;
        if (request.ready || request.done || request.cancelled || request.observeOnly) {
            ToastStore.sharedInstance().dismissToast(this.props.toastKey);
        }
    };

    cancel = () => {
        ToastStore.sharedInstance().dismissToast(this.props.toastKey);
        try {
            this.props.request.cancel();
        } catch (err) {
            console.error("Error while cancelling verification request", err);
        }
    }

    accept = async () => {
        ToastStore.sharedInstance().dismissToast(this.props.toastKey);
        const {request} = this.props;
        // no room id for to_device requests
        const cli = MatrixClientPeg.get();
        try {
            await request.accept();
            if (request.channel.roomId) {
                dis.dispatch({
                    action: 'view_room',
                    room_id: request.channel.roomId,
                    should_peek: false,
                });
                dis.dispatch({
                    action: "set_right_panel_phase",
                    phase: RIGHT_PANEL_PHASES.EncryptionPanel,
                    refireParams: {
                        verificationRequest: request,
                        member: cli.getUser(request.otherUserId),
                    },
                });
            } else {
                const VerificationRequestDialog = sdk.getComponent("views.dialogs.VerificationRequestDialog");
                Modal.createTrackedDialog('Incoming Verification', '', VerificationRequestDialog, {
                    verificationRequest: request,
                }, null, /* priority = */ false, /* static = */ true);
            }
        } catch (err) {
            console.error(err.message);
        }
    };

    render() {
        const FormButton = sdk.getComponent("elements.FormButton");
        const {request} = this.props;
        const userId = request.otherUserId;
        const roomId = request.channel.roomId;
        let nameLabel = roomId ? userLabelForEventRoom(userId, roomId) : userId;
        // for legacy to_device verification requests
        if (nameLabel === userId) {
            const client = MatrixClientPeg.get();
            const user = client.getUser(userId);
            if (user && user.displayName) {
                nameLabel = _t("%(name)s (%(userId)s)", {name: user.displayName, userId});
            }
        }
        const declineLabel = this.state.counter == 0 ?
            _t("Decline") :
            _t("Decline (%(counter)s)", {counter: this.state.counter});
        return (<div>
            <div className="mx_Toast_description">{nameLabel}</div>
            <div className="mx_Toast_buttons" aria-live="off">
                <FormButton label={declineLabel} kind="danger" onClick={this.cancel} />
                <FormButton label={_t("Accept")} onClick={this.accept} />
            </div>
        </div>);
    }
}

VerificationRequestToast.propTypes = {
    request: PropTypes.object.isRequired,
    toastKey: PropTypes.string.isRequired,
};
