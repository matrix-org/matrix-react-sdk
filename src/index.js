/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
