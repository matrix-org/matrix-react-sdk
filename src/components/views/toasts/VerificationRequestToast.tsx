/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";

import * as sdk from "../../../index";
import { _t } from '../../../languageHandler';
import {MatrixClientPeg} from '../../../MatrixClientPeg';
import {RIGHT_PANEL_PHASES} from "../../../stores/RightPanelStorePhases";
import {userLabelForEventRoom} from "../../../utils/KeyVerificationStateObserver";
import dis from "../../../dispatcher/dispatcher";
import ToastStore from "../../../stores/ToastStore";
import Modal from "../../../Modal";
import GenericToast from "./GenericToast";
import {VerificationRequest} from "matrix-js-sdk/src/crypto/verification/request/VerificationRequest";
import {DeviceInfo} from "matrix-js-sdk/src/crypto/deviceinfo";

interface IProps {
    toastKey: string;
    request: VerificationRequest;
}

interface IState {
    counter: number;
    device?: DeviceInfo;
}

export default class VerificationRequestToast extends React.PureComponent<IProps, IState> {
    private intervalHandle: NodeJS.Timeout;

    constructor(props) {
        super(props);
        this.state = {counter: Math.ceil(props.request.timeout / 1000)};
    }

    async componentDidMount() {
        const {request} = this.props;
        if (request.timeout && request.timeout > 0) {
            this.intervalHandle = setInterval(() => {
                let {counter} = this.state;
                counter = Math.max(0, counter - 1);
                this.setState({counter});
            }, 1000);
        }
        request.on("change", this._checkRequestIsPending);
        // We should probably have a separate class managing the active verification toasts,
        // rather than monitoring this in the toast component itself, since we'll get problems
        // like the toasdt not going away when the verification is cancelled unless it's the
        // one on the top (ie. the one that's mounted).
        // As a quick & dirty fix, check the toast is still relevant when it mounts (this prevents
        // a toast hanging around after logging in if you did a verification as part of login).
        this._checkRequestIsPending();

        if (request.isSelfVerification) {
            const cli = MatrixClientPeg.get();
            this.setState({device: cli.getStoredDevice(cli.getUserId(), request.channel.deviceId)});
        }
    }

    componentWillUnmount() {
        clearInterval(this.intervalHandle);
        const {request} = this.props;
        request.off("change", this._checkRequestIsPending);
    }

    _checkRequestIsPending = () => {
        const {request} = this.props;
        if (!request.canAccept) {
            ToastStore.sharedInstance().dismissToast(this.props.toastKey);
        }
    };

    cancel = () => {
        ToastStore.sharedInstance().dismissToast(this.props.toastKey);
        try {
            this.props.request.cancel();
        } catch (err) {
            console.error("Error while cancelling verification request", err);
        }
    }

    accept = async () => {
        ToastStore.sharedInstance().dismissToast(this.props.toastKey);
        const {request} = this.props;
        // no room id for to_device requests
        const cli = MatrixClientPeg.get();
        try {
            if (request.channel.roomId) {
                dis.dispatch({
                    action: 'view_room',
                    room_id: request.channel.roomId,
                    should_peek: false,
                });
                dis.dispatch({
                    action: "set_right_panel_phase",
                    phase: RIGHT_PANEL_PHASES.EncryptionPanel,
                    refireParams: {
                        verificationRequest: request,
                        member: cli.getUser(request.otherUserId),
                    },
                });
            } else {
                const VerificationRequestDialog = sdk.getComponent("views.dialogs.VerificationRequestDialog");
                Modal.createTrackedDialog('Incoming Verification', '', VerificationRequestDialog, {
                    verificationRequest: request,
                }, null, /* priority = */ false, /* static = */ true);
            }
            await request.accept();
        } catch (err) {
            console.error(err.message);
        }
    };

    render() {
        const {request} = this.props;
        let nameLabel;
        if (request.isSelfVerification) {
            if (this.state.device) {
                nameLabel = _t("From %(deviceName)s (%(deviceId)s)", {
                    deviceName: this.state.device.getDisplayName(),
                    deviceId: this.state.device.deviceId,
                });
            }
        } else {
            const userId = request.otherUserId;
            const roomId = request.channel.roomId;
            nameLabel = roomId ? userLabelForEventRoom(userId, roomId) : userId;
            // for legacy to_device verification requests
            if (nameLabel === userId) {
                const client = MatrixClientPeg.get();
                const user = client.getUser(userId);
                if (user && user.displayName) {
                    nameLabel = _t("%(name)s (%(userId)s)", {name: user.displayName, userId});
                }
            }
        }
        const declineLabel = this.state.counter === 0 ?
            _t("Decline") :
            _t("Decline (%(counter)s)", {counter: this.state.counter});

        return <GenericToast
            description={nameLabel}
            acceptLabel={_t("Accept")}
            onAccept={this.accept}
            rejectLabel={declineLabel}
            onReject={this.cancel}
        />;
    }
}
