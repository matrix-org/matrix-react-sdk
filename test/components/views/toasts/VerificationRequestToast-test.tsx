/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { ComponentProps } from "react";
import { Mocked } from "jest-mock";
import { act, render, RenderResult } from "@testing-library/react";
import {
    VerificationRequest,
    VerificationRequestEvent,
} from "matrix-js-sdk/src/crypto/verification/request/VerificationRequest";
import { DeviceInfo } from "matrix-js-sdk/src/crypto/deviceinfo";
import { IMyDevice, MatrixClient } from "matrix-js-sdk/src/client";
import { TypedEventEmitter } from "matrix-js-sdk/src/models/typed-event-emitter";

import VerificationRequestToast from "../../../../src/components/views/toasts/VerificationRequestToast";
import { flushPromises, getMockClientWithEventEmitter, mockClientMethodsUser } from "../../../test-utils";

function renderComponent(
    props: Partial<ComponentProps<typeof VerificationRequestToast>> & { request: VerificationRequest },
): RenderResult {
    const propsWithDefaults = {
        toastKey: "test",
        ...props,
    };

    return render(<VerificationRequestToast {...propsWithDefaults} />);
}

describe("VerificationRequestToast", () => {
    let client: Mocked<MatrixClient>;

    beforeEach(() => {
        client = getMockClientWithEventEmitter({
            ...mockClientMethodsUser(),
            getStoredDevice: jest.fn(),
            getDevice: jest.fn(),
        });
    });

    it("should render a self-verification", async () => {
        const otherDeviceId = "other_device";
        const otherIDevice: IMyDevice = { device_id: otherDeviceId, last_seen_ip: "1.1.1.1" };
        client.getDevice.mockResolvedValue(otherIDevice);

        const otherDeviceInfo = new DeviceInfo(otherDeviceId);
        otherDeviceInfo.unsigned = { device_display_name: "my other device" };
        client.getStoredDevice.mockReturnValue(otherDeviceInfo);

        const request = makeMockVerificationRequest({
            isSelfVerification: true,
            otherDeviceId,
        });
        const result = renderComponent({ request });
        await act(async () => {
            await flushPromises();
        });
        expect(result.container).toMatchSnapshot();
    });

    it("should render a cross-user verification", async () => {
        const otherUserId = "@other:user";
        const request = makeMockVerificationRequest({
            isSelfVerification: false,
            otherUserId,
        });
        const result = renderComponent({ request });
        await act(async () => {
            await flushPromises();
        });
        expect(result.container).toMatchSnapshot();
    });
});

function makeMockVerificationRequest(props: Partial<VerificationRequest> = {}): Mocked<VerificationRequest> {
    const request = new TypedEventEmitter<VerificationRequestEvent, any>();
    Object.assign(request, {
        ...props,
    });
    return request as unknown as Mocked<VerificationRequest>;
}
