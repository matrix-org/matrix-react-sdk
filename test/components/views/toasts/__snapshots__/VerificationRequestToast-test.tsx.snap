// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerificationRequestToast should render a cross-user verification 1`] = `
<div>
  <div>
    <div
      class="mx_Toast_description"
    >
      @alice:domain (@other:user)
    </div>
    <div
      aria-live="off"
      class="mx_Toast_buttons"
    >
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_danger_outline"
        role="button"
        tabindex="0"
      >
        Ignore
      </div>
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
        role="button"
        tabindex="0"
      >
        Verify Session
      </div>
    </div>
  </div>
</div>
`;

exports[`VerificationRequestToast should render a self-verification 1`] = `
<div>
  <div>
    <div
      class="mx_Toast_description"
    >
      my other device
      <div
        class="mx_Toast_detail"
      >
        other_device from 1.1.1.1
      </div>
    </div>
    <div
      aria-live="off"
      class="mx_Toast_buttons"
    >
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_danger_outline"
        role="button"
        tabindex="0"
      >
        Ignore
      </div>
      <div
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
        role="button"
        tabindex="0"
      >
        Verify Session
      </div>
    </div>
  </div>
</div>
`;
